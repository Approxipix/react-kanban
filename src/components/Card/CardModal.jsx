import React, { Component } from 'react';
import styled from 'styled-components';
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import CardTitleEdit from './CardTitleEdit';
import CardDescEdit from './CardDescEdit';

const Backdrop = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.3);
  z-index: 100;
`;

const Container = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  width: 640px;
  max-width: 100%;
  max-height: calc(100% - 8rem);
  padding: 1rem;
  border-radius: .2rem;
  background-color: #fff;
  transform: translate(-50%, -50%);
`;

const CloseButton = styled.button`
  position: absolute;
  top: 1rem;
  right: 1rem;
  font-size: 1.3rem;
  z-index: 110;
  cursor: pointer;
`;

const Wrapper = styled.div`
  position: relative;
  margin-bottom: 1rem;
  padding-left: 1.8rem;
`;

const Title = styled.h4`
  padding: .3rem .5rem;
  font-size: 1.5rem;
  font-weight: 600;
`;

const Icon = styled.div`
  position: absolute;
  left: 0;
  top: .6rem;
  font-size: 1rem;
`;


const Col = styled.div`

`;



const SubTitle = styled.h4`

`;

const Actions = styled.h4`

`;

const Button = styled.h4`

`;

const ButtonIcon = styled.h4`

`;

const Description = styled.p`
  padding: .3rem .5rem;
`;

const Close = styled.button`
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  background: transparent;
  border: none;
  outline: none;
`;


class CardModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isTitleEditing: false,
      isDescEditing: false,
    }
  }

  toggleIsTitleEditing = () => {
    this.setState({
      isTitleEditing: !this.state.isTitleEditing
    })
  };

  toggleIsDescEditing = () => {
    this.setState({
      isDescEditing: !this.state.isDescEditing
    })
  };

  componentDidMount() {
    document.documentElement.classList.add('scroll-disabled');
  }

  componentWillUnmount() {
    document.documentElement.classList.remove('scroll-disabled');
  }


  render() {
    const { card, cardIndex, listIndex } = this.props;
    const { isTitleEditing, isDescEditing } = this.state;
    if (!card) return null;
    return (
      <Backdrop>
        <Close onClick={() => this.props.toggleModal()} />
        <Container onClick={(e) => e.preventDefault()}>
          <CloseButton onClick={() => this.props.toggleModal()}>
            <FontAwesomeIcon icon="times" />
          </CloseButton>
          <Wrapper>
            {!isTitleEditing ? (
              <Title onClick={() => this.toggleIsTitleEditing()}>
                {card.title}
              </Title>
            ) : (
              <CardTitleEdit
                card={card}
                cardIndex={cardIndex}
                listIndex={listIndex}
                toggleIsTitleEditing={this.toggleIsTitleEditing}
              />
            )}
            <Icon>
              <FontAwesomeIcon icon="window-maximize" />
            </Icon>
          </Wrapper>
          <Wrapper>
            <Title>
              Description
            </Title>
            {!isDescEditing ? (
              <Description onClick={() => this.toggleIsDescEditing()}>
                {card.description}
              </Description>
            ) : (
              <CardDescEdit
                card={card}
                cardIndex={cardIndex}
                listIndex={listIndex}
                toggleIsDescEditing={this.toggleIsDescEditing}
              />
            )}
            <Icon>
              <FontAwesomeIcon icon="align-left" />
            </Icon>
          </Wrapper>

                </ButtonIcon>
              </Button>
            </Actions>
          </Col>
        </Container>
      </Backdrop>
    )
  }
}

function mapStateToProps(state, ownProps) {
  const { rootReducer } = state;
  const board = rootReducer.boards[rootReducer.currentBoardIndex];
  const list = board.lists[1];
  // const list = board.lists[ownProps.listIndex];
  const card = list.cards[1];
  // const card = list.cards[ownProps.cardIndex];
  return {
    board: board,
    list: list,
    card: card,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({

    }, dispatch)
  };
}


export default connect(mapStateToProps, mapDispatchToProps)(CardModal);
