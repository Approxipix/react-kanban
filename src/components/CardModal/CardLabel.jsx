import React, { Component } from 'react';
import styled from 'styled-components';

const LabelList = styled.ul`
  display: flex;
`;

const LabelItem = styled.li`
  height: 2rem;
  width: 2rem;
  margin-right: .5rem;
  background-color: ${props => props.value};
  cursor: pointer;
  border-radius: .2rem;
`;


class CardLabel extends Component {
  render() {
    const { cardLabels } = this.props;
    return (
      <LabelList>
        {cardLabels.map((label, index) => (
          <LabelItem key={index} value={label.color}/>
        ))}
      </LabelList>
    )
  }
}



export default CardLabel;
