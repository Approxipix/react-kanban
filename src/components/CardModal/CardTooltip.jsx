import React, { Component } from 'react';
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import ClickOutside from "../ClickOutside/ClickOutside";
import styled from 'styled-components';

const Wrapper = styled.div`
  position: absolute;
  top: 2.5rem;
  left: 0;
  width: 17rem;
  background-color: #F4F5F7;
  border-radius: .2rem;
  box-shadow: rgba(0, 0, 0, 0.3) .125rem .125rem .5rem;
  z-index: 10;
  @media (max-width: 910px) {
    right: 0;
    left: auto;
  }
`;

const Header = styled.div`
  position: relative;
  padding: .5rem;
  border-bottom: 1px solid rgba(9,30,66,.13);
`;

const Title = styled.h4`
  font-size: 1rem;
  color: #6b778c;
  text-align: center;
`;

const Button = styled.button`
  position: absolute;
  top: .5rem;
  right: .5rem;
  font-size: 1rem;
  color: #6b778c;
  transition: color .1s ease-in;
  &:hover {
    color: #42526e;
  }
`;
Button.displayName = 'Button';

const Body = styled.div`
   padding: .5rem;
`;

class CardTooltip extends Component {
  componentDidMount() {
    document.addEventListener('keydown', this.outerClick);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.outerClick);
  }

  outerClick = (e) => {
    const { isOpened } = this.props;
    if (!isOpened) return null;
    if (e.keyCode === 27) {
      this.props.toggleTooltip(null);
    }
    return null;
  };

  render() {
    const { title, body, isOpened } = this.props;
    if (!isOpened) return null;
    return (
      <ClickOutside toggleOpened={this.props.toggleTooltip}>
        <Wrapper>
          <Header>
            <Title>
              {title}
            </Title>
            <Button onClick={() => this.props.toggleTooltip()}>
              <FontAwesomeIcon icon="times"/>
            </Button>
          </Header>
          <Body>
            {body}
          </Body>
        </Wrapper>
      </ClickOutside>
    )
  }
}

CardTooltip.defaultProps = {
  title: '',
  isOpened: false,
  toggleTooltip: () => {},
};

CardTooltip.propTypes = {
  isOpened: PropTypes.bool.isRequired,
  title: PropTypes.string.isRequired,
  toggleTooltip: PropTypes.func.isRequired,
  body: PropTypes.element.isRequired,
};

export default CardTooltip;
