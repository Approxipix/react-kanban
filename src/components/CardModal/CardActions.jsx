import React, { Component } from 'react';
import CardChecklistAdd from '../Checklist/CheckListAdd';
import CardDelete from '../Card/CardDelete';
import LabelAdd from '../Label/LabelAdd';
import CardTooltip from './CardTooltip';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import styled from 'styled-components';
import PropTypes from "prop-types";

const Wrapper = styled.div`
  position: relative;
  padding: .5rem 0;
  :not(:last-child) {
    margin-bottom: 1rem;
  }
`;

const Title = styled.h4`
  margin-bottom: .5rem;
  font-size: .75rem;
  color: #6b778c;
  font-weight: 500;
  text-transform: uppercase;
`;

const ActionList = styled.ul`
  display: flex;
  flex-direction: column;
`;
ActionList.displayName = 'ActionList';

const ActionItem = styled.li`
  position: relative;
  &:not(:last-child) {
    margin-bottom: .5rem;
  }
`;

const Link = styled.a`
  display: flex;
  align-items: center;
  width: 100%;
  padding: .5rem .75rem;
  font-size: .875rem;
  font-weight: 600;
  border-radius: .2rem;
  text-align: left;
  background-color: #ebecf0;
  box-shadow: 0 1px 0 0 rgba(9, 30, 66, .13);
  transition: all .2s ease-in;
  cursor: pointer;
  &:hover {
    background-color: #dfe1e6;
    box-shadow: 0 1px 0 0 rgba(9, 30, 66, .25);
  }
`;
Link.displayName = 'Link';

const Icon = styled.div`
  margin-right: .5rem;
`;

class CardActions extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: null
    }
  }

  toggleIsOpened = (key) => {
    this.setState({
      isOpened: key !== this.state.isOpened ? key : null
    })
  };

  render() {
    const { listId, card, title, actions } = this.props;
    const { isOpened } = this.state;
    let actionItems = [
      {
        title: 'Label',
        icon: 'tag',
        component: (
          <LabelAdd
            cardId={card._cardId}
            cardLabels={card.cardLabels}
          />
        )
      },
      {
        title: 'CheckList',
        icon: 'check-square',
        component: (
          <CardChecklistAdd
            cardId={card._cardId}
            toggleTooltip={() => this.toggleIsOpened('CheckList')}
          />
        )
      },
      {
        title: 'Delete',
        icon: 'trash',
        component: (
          <CardDelete
            cardId={card._cardId}
            listId={listId}
            toggleTooltip={() => this.toggleIsOpened('Delete')}
          />
        )
      },
    ];
    actionItems = actionItems.filter(item => actions.some(title => title === item.title));

    return (
      <Wrapper>
        <Title>
          {title}
        </Title>
        <ActionList>
          {actionItems.map((action, index) => (
            <ActionItem key={index}>
              <Link onClick={() => this.toggleIsOpened(action.title)}>
                <Icon>
                  <FontAwesomeIcon icon={action.icon} />
                </Icon>
                {action.title}
              </Link>
              <CardTooltip
                title={action.title}
                isOpened={isOpened === action.title}
                toggleTooltip={() => this.toggleIsOpened(action.title)}
                body={action.component}
              />
            </ActionItem>
          ))}
        </ActionList>
      </Wrapper>
    )
  }
}

CardActions.defaultProps = {
  card: {},
  listId: '',
  title: '',
  actions: [],
};

CardActions.propTypes = {
  card: PropTypes.shape({
    _cardId: PropTypes.string.isRequired,
    title: PropTypes.string,
    description: PropTypes.string,
    checkLists: PropTypes.arrayOf(PropTypes.string),
    cardLabels: PropTypes.array,
  }).isRequired,
  listId: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  actions: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default CardActions;
