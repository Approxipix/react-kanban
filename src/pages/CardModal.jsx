import React from 'react';
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { history } from "../redux/store";
import { Redirect } from "react-router";
import CardActions from "../components/CardModal/CardActions";
import CheckList from "../components/Checklist/CheckList";
import CardTitle from "../components/CardModal/CardTitle/CardTitle";
import CardDescription from "../components/CardModal/CardDescription/CardDescription";
import Label from "../components/Label/Label";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import styled from 'styled-components';

const Backdrop = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(0, 0, 0, 0.5);
  z-index: 100;
`;

const CloseBackdrop = styled.button`
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  opacity: 0;
  background: transparent;
  border: none;
  outline: none;
`;

const Wrapper = styled.div`
  position: absolute;
  top: 5rem;
  left: 50%;
  max-width: 45rem;
  width: 100%;
  max-height: calc(100% - 8rem);
  padding: 1rem;
  border-radius: .2rem;
  background-color: #f4f5f7;
  transform: translateX(-50%);
  @media (max-width: 770px) {
    left: 1rem;
    right: 1rem;
    width: auto;
    transform: translateX(0);
  }
`;

const CloseButton = styled.button`
  position: absolute;
  top: 1rem;
  right: 1rem;
  color: #6b778c;
  font-size: 1.3rem;
  transition: color .1s ease-in;
  z-index: 110;
  &:hover {
    color: #42526e;
  }
`;

const Container = styled.div`
  position: relative;
  &:not(:last-child) {
    margin-bottom: 1rem;
  }  
`;

const Header = styled.div`
  position: relative;
  margin-bottom: .3rem;
  padding-left: calc(2rem - .5rem);
`;

const Body = styled.div`
  padding-left: 2rem;
`;

const Title = styled.h3`
  padding: .3rem .5rem;
  font-size: 1.2rem;
  font-weight: 600;
`;

const Icon = styled.div`
  position: absolute;
  left: 0;
  top: 50%;
  font-size: 1rem;
  transform: translateY(-50%);
`;

const Row = styled.div`
  display: grid;
  grid-template-columns: 3fr 1fr;
  grid-gap: 1rem;
  @media (max-width: 600px) {
    grid-template-columns: 1fr;
  }
`;

const Col = styled.div`
`;

const TitleContainer = (props) => {
  const { card } = props;
  return (
    <Container>
      <Header>
        <Icon>
          <FontAwesomeIcon icon="window-maximize" />
        </Icon>
        <CardTitle
          cardId={card._cardId}
          cardTitle={card.title}
        />
      </Header>
    </Container>
  )
};

const DescriptionContainer = (props) => {
  const { card } = props;
  return (
    <Container>
      <Header>
        <Icon>
          <FontAwesomeIcon icon="align-left" />
        </Icon>
        <Title>
          Description
        </Title>
      </Header>
      <Body>
        <CardDescription
          cardId={card._cardId}
          cardDescription={card.description}
        />
      </Body>
    </Container>
  )
};

const LabelContainer = (props) => {
  const { card } = props;
  if (card.cardLabels.length === 0) return null;
  return (
    <Container>
      <Header>
        <Icon>
          <FontAwesomeIcon icon="tag" />
        </Icon>
        <Title>
          Labels
        </Title>
      </Header>
      <Body>
        <Label cardLabels={card.cardLabels}/>
      </Body>
    </Container>
  )
};

const ChecklistContainer = (props) => {
  const { card } = props;
  if (card.checkLists.length === 0) return null;
  return (
    card.checkLists.map((checkListId, index) => (
      <Container key={index}>
        <CheckList
          cardId={card._cardId}
          checkListId={checkListId}
        />
      </Container>
    ))
  )
};

const CardModal = (props) => {
  const { card, lists, currentBoardId } = props;
  if (!currentBoardId) return null;
  const boardUrl = `/b/${currentBoardId}`;
  if (!card) return <Redirect to={boardUrl} />;

  let listId = '';
  Object.keys(lists).forEach(list => {
    lists[list].cards.forEach(item => {
      if (item === card._cardId) {
        listId = list;
      }
    })
  });
  return (
    <Backdrop>
      <CloseBackdrop onClick={() => history.push(boardUrl)} />
      <Wrapper onClick={(e) => e.preventDefault()}>
        <CloseButton onClick={() => history.push(boardUrl)}>
          <FontAwesomeIcon icon="times" />
        </CloseButton>
        <TitleContainer card={card}/>
        <Row>
          <Col>
            <DescriptionContainer card={card}/>
            <LabelContainer card={card}/>
            <ChecklistContainer card={card} />
          </Col>
          <Col>
            <CardActions
              card={card}
              title="Add to card"
              actions={["Label", "CheckList"]}
              listId={listId}
            />
            <CardActions
              card={card}
              title="Actions"
              actions={["Delete"]}
              listId={listId}
            />
          </Col>
        </Row>
      </Wrapper>
    </Backdrop>
  )
};

CardModal.propTypes = {
  currentBoardId: PropTypes.string,
  card: PropTypes.shape({
    _cardId: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    checkLists: PropTypes.arrayOf(PropTypes.string).isRequired,
    cardLabels: PropTypes.array.isRequired,
  }),
  lists: PropTypes.objectOf(
    PropTypes.shape({
      _listId: PropTypes.string.isRequired,
      title: PropTypes.string,
      cards: PropTypes.arrayOf(PropTypes.string).isRequired
    })
  ).isRequired,
};

function mapStateToProps(state, ownProps) {
  return {
    currentBoardId: state.rootReducer.currentBoardID,
    card: state.cardReducer[ownProps.match.params.cardId],
    lists: state.listReducer,
  }
}

export default connect(mapStateToProps)(CardModal);
