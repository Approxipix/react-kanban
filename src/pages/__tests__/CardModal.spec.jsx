import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CardModal from '../CardModal';

const mockStore = configureStore([]);

describe('<CardModal>', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      rootReducer: {
        currentBoardID: 'BoardID',
      },
      listReducer: {
        ListID: {
          _listId: 'ListID',
          title: 'List Title',
          cards: ['Card1']
        }
      },
      cardReducer: {
        Card1: {
          _cardId: 'Card1ID',
          title: 'Card1 Title',
          description: 'Card1 Description',
          cardLabels: ['CardLabel1', 'CardLabel2'],
          checkLists: ['CheckList1', 'CheckList2', 'CheckList3'],
        },
      }
    });
    const props = {
      match: { params: { cardId: 'Card1' } }
    };

    const wrapper = shallow(<CardModal store={store} {...props} />,);
    component = wrapper.find('CardModal').dive();
  });

  it('should redirect to `/b/BoardID` if the card is not found', () => {
    const props = {
      match: { params: { cardId: '' } }
    };
    const wrapper = shallow(<CardModal store={store} {...props} />,);
    const component = wrapper.find('CardModal').dive();

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with given state from Redux store and correct props', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
