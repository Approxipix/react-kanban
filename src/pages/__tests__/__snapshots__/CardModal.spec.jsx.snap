// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardModal> should redirect to \`/b/BoardID\` if the card is not found 1`] = `
<Redirect
  to="/b/BoardID"
/>
`;

exports[`<CardModal> should render with given state from Redux store and correct props 1`] = `
<styled.div>
  <styled.button
    onClick={[Function]}
  />
  <styled.div
    onClick={[Function]}
  >
    <styled.button
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="times"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </styled.button>
    <TitleContainer
      card={
        Object {
          "_cardId": "Card1ID",
          "cardLabels": Array [
            "CardLabel1",
            "CardLabel2",
          ],
          "checkLists": Array [
            "CheckList1",
            "CheckList2",
            "CheckList3",
          ],
          "description": "Card1 Description",
          "title": "Card1 Title",
        }
      }
    />
    <styled.div>
      <styled.div>
        <DescriptionContainer
          card={
            Object {
              "_cardId": "Card1ID",
              "cardLabels": Array [
                "CardLabel1",
                "CardLabel2",
              ],
              "checkLists": Array [
                "CheckList1",
                "CheckList2",
                "CheckList3",
              ],
              "description": "Card1 Description",
              "title": "Card1 Title",
            }
          }
        />
        <LabelContainer
          card={
            Object {
              "_cardId": "Card1ID",
              "cardLabels": Array [
                "CardLabel1",
                "CardLabel2",
              ],
              "checkLists": Array [
                "CheckList1",
                "CheckList2",
                "CheckList3",
              ],
              "description": "Card1 Description",
              "title": "Card1 Title",
            }
          }
        />
        <ChecklistContainer
          card={
            Object {
              "_cardId": "Card1ID",
              "cardLabels": Array [
                "CardLabel1",
                "CardLabel2",
              ],
              "checkLists": Array [
                "CheckList1",
                "CheckList2",
                "CheckList3",
              ],
              "description": "Card1 Description",
              "title": "Card1 Title",
            }
          }
        />
      </styled.div>
      <styled.div>
        <CardActions
          actions={
            Array [
              "Label",
              "CheckList",
            ]
          }
          card={
            Object {
              "_cardId": "Card1ID",
              "cardLabels": Array [
                "CardLabel1",
                "CardLabel2",
              ],
              "checkLists": Array [
                "CheckList1",
                "CheckList2",
                "CheckList3",
              ],
              "description": "Card1 Description",
              "title": "Card1 Title",
            }
          }
          listId=""
          title="Add to card"
        />
        <CardActions
          actions={
            Array [
              "Delete",
            ]
          }
          card={
            Object {
              "_cardId": "Card1ID",
              "cardLabels": Array [
                "CardLabel1",
                "CardLabel2",
              ],
              "checkLists": Array [
                "CheckList1",
                "CheckList2",
                "CheckList3",
              ],
              "description": "Card1 Description",
              "title": "Card1 Title",
            }
          }
          listId=""
          title="Actions"
        />
      </styled.div>
    </styled.div>
  </styled.div>
</styled.div>
`;
